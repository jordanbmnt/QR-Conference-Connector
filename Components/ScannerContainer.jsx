import { useState, useEffect } from "react";
import { StyleSheet } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import {
  setAlertTheme,
  setAlertVisibility,
  setContactInfo,
  setPreviousPage,
  setScanViewVisible,
  setViewContact,
} from "../Redux/appSlice";
import * as SQLite from "expo-sqlite";
import { BarCodeScanner } from "expo-barcode-scanner";
import { Box } from "@gluestack-ui/themed";
import { infoParser } from "../HelperFunctions/infoParser/infoParser";
import {
  addContact,
  createTable,
  getAllContactsAsc,
} from "../HelperFunctions/sqlHelperFunctions";
import ContactList from "./ContactList";
import PermissionLoader from "./PermissionLoader/PermissionLoader";
import Scanner from "./Scanner";
import ImageStatus from "./ImageStatus/ImageStatus";
import NavButtons from "./NavButtons";

export default ScannerContainer = () => {
  const dispatch = useDispatch();
  const scanViewVisible = useSelector(
    (state) => state.appSlice.scanViewVisible
  );
  const contactInfo = useSelector((state) => state.appSlice.contactInfo);
  const db = SQLite.openDatabase("../contacts.db");
  const [isValidQR, setIsValidQR] = useState(true);
  const [isLoading, setIsLoading] = useState(true);
  const [hasPermission, setHasPermission] = useState(null);
  const [tempContactList, setTempContactList] = useState([]);

  useEffect(() => {
    const getBarCodeScannerPermissions = async () => {
      const { status } = await BarCodeScanner.requestPermissionsAsync();
      setHasPermission(status === "granted");
    };
    getBarCodeScannerPermissions();
  }, []);

  useEffect(() => {
    createTable(db);
    setIsLoading(false);
  }, []);

  useEffect(() => {
    getAllContactsAsc(db).then(
      (results) => {
        setTempContactList(results);
      },
      (__) => {
        alertFunction("error");
      }
    );
  }, [handleBarCodeScanned, setIsLoading]);

  if (hasPermission === false) {
    return <ImageStatus theme={"permission"} />;
  }

  if (isLoading || hasPermission === null) {
    return <PermissionLoader />;
  }

  const alertFunction = (alertMsg) => {
    dispatch(setAlertTheme(alertMsg));
    dispatch(setAlertVisibility(true));
    setTimeout(() => {
      dispatch(setAlertVisibility(false));
      dispatch(setAlertTheme(""));
    }, 2600);
  };

  const updateTempContactList = (info) => {
    tempContactList.length > 0
      ? setTempContactList(
          [...tempContactList, info].sort((a, b) =>
            a.name.localeCompare(b.name)
          )
        )
      : setTempContactList([info]);
    alertFunction("success");
  };

  const handleBarCodeScanned = ({ __, data }) => {
    if (!data || typeof data !== "string" || data.split("\n").length < 6) {
      setIsValidQR(false);
      setTimeout(() => {
        setIsValidQR(true);
      }, 2550);
      return;
    }
    dispatch(setScanViewVisible(false));
    const info = infoParser(data.split("\n"));

    const contactDoesExist = tempContactList.filter((contact) => {
      if (contact.email === info.email) return true;
    }).length;

    contactDoesExist === 0
      ? addContact({ database: db, contactDetails: info }).then(
          (results) => {
            results && updateTempContactList(info);
          },
          (__) => {
            alertFunction("errorUpload");
          }
        )
      : alertFunction("duplicate");
    dispatch(setContactInfo(info));
    dispatch(setPreviousPage(true));
    dispatch(setViewContact(true));
  };

  return (
    <Box
      height='60%'
      rounded='$lg'
      sx={styles.container}
      justifyContent='space-between'
      alignItems='center'
    >
      {!isValidQR && <ImageStatus theme={"invalidQR"} />}

      {scanViewVisible && isValidQR && (
        <Scanner
          scanCheck={!scanViewVisible}
          handleScanned={handleBarCodeScanned}
        />
      )}

      {!scanViewVisible &&
        (tempContactList.length === 0 &&
        Object.values(contactInfo).length === 0 ? (
          <ImageStatus theme={"noContacts"} />
        ) : (
          <ContactList contacts={tempContactList} />
        ))}

      <NavButtons validity={isValidQR} />
    </Box>
  );
};

const styles = StyleSheet.create({
  container: {
    "@base": {
      my: "$16",
      mx: "$5",
      height: "80%",
    },
    "@lg": {
      my: "$24",
      mx: "$32",
    },
    borderWidth: "$2",
    borderColor: "$orange100",
  },
  button: {
    marginTop: 0,
    margin: 30,
    marginBottom: 40,
  },
});
