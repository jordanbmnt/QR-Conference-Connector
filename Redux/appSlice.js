import { createSlice } from '@reduxjs/toolkit';

export const appSlice = createSlice({
  name: 'appSlice',
  initialState: {
    viewContactInfo: false,
    scanViewVisible: true,
    previousPage: false,
    contactInfo: {},
    alertInfo: {
      visibility: false,
      theme: "",
    }
  },
  reducers: {
    setViewContact: (state, action) => {
      state.viewContactInfo = action.payload;
    },
    setScanViewVisible: (state, action) => {
      state.scanViewVisible = action.payload;
    },
    setPreviousPage: (state, action) => {
      state.previousPage = action.payload;
    },
    setContactInfo: (state, action) => {
      state.contactInfo = action.payload;
    },
    setAlertVisibility: (state, action) => {
      state.alertInfo.visibility = action.payload;
    },
    setAlertTheme: (state, action) => {
      state.alertInfo.theme = action.payload;
    },
  },
});

export const { setViewContact, setContactInfo, setScanViewVisible, setPreviousPage, setAlertTheme, setAlertVisibility } = appSlice.actions;


export default appSlice.reducer;
