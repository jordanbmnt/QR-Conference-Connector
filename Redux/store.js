import { configureStore } from '@reduxjs/toolkit';
import rootReducer from './reducers'

export const setupUserInfoStore = (preloadedState) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const store = configureStore({
  reducer: rootReducer
})