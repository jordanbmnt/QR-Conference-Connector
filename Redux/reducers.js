import { combineReducers } from 'redux';
import  appSlice  from './appSlice';

export default combineReducers({ appSlice })