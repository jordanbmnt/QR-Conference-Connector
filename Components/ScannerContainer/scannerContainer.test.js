import ScannerContainer from "./ScannerContainer";
import renderer, { act } from "react-test-renderer";
import { Provider } from "react-redux";
import { GluestackUIStyledProvider } from "@gluestack-ui/themed";
import { config } from "@gluestack-ui/config";
import { setupUserInfoStore } from "../../Redux/store";

function MockBarCodeScanner() {
  return <></>;
}

MockBarCodeScanner.requestPermissionsAsync = async function () {
  return { status: "granted" };
};

jest.mock("expo-barcode-scanner", () => ({
  __esModule: true,
  BarCodeScanner: MockBarCodeScanner,
}));

jest.mock("expo-sqlite", () => ({
  openDatabase: jest.fn(() => ({
    transaction: jest.fn(),
  })),
}));

describe("ScannerContainer", function () {
  let scannerContainer;
  const initialState = {
    appSlice: {
      scanViewVisible: true,
      contactInfo: {},
    },
  };

  beforeAll(async () => {
    const store = setupUserInfoStore(initialState);
    await act(async () => {
      scannerContainer = renderer.create(
        <Provider store={store}>
          <GluestackUIStyledProvider config={config}>
            <ScannerContainer />
          </GluestackUIStyledProvider>
        </Provider>
      );
    });
  });

  it("should render scanner component correctly", () => {
    act(() => {
      expect(scannerContainer).toMatchSnapshot();
    });
  });

  it("should display a barcodeScanner", function () {
    expect(scannerContainer.toJSON().children.length).toBe(2);
  });
});
