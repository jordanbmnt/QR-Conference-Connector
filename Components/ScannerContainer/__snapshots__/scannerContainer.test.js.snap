// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScannerContainer should render scanner component correctly 1`] = `
<View
  data-style=""
  dataSet={
    {
      "style": "",
    }
  }
  style={
    [
      undefined,
      {},
      {},
      {},
      {
        "alignItems": "center",
        "borderColor": "#ffedd5",
        "borderRadius": 8,
        "borderWidth": 2,
        "height": "60%",
        "justifyContent": "space-between",
      },
      {
        "height": "80%",
        "marginHorizontal": 20,
        "marginVertical": 64,
      },
      undefined,
    ]
  }
  sx={
    {
      "@base": {
        "height": "80%",
        "mx": "$5",
        "my": "$16",
      },
      "@lg": {
        "mx": "$32",
        "my": "$24",
      },
      "borderColor": "$orange100",
      "borderWidth": "$2",
    }
  }
>
  <View
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        {},
        {},
        {
          "height": "90%",
          "paddingBottom": 0,
          "width": "100%",
        },
        undefined,
      ]
    }
    sx={
      {
        "height": "90%",
        "paddingBottom": 0,
        "width": "100%",
      }
    }
  >
    <Text
      accessibilityRole="header"
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        {
          "backgroundColor": "#131721",
          "borderBottomLeftRadius": 0,
          "borderBottomRightRadius": 0,
          "borderRadius": 9,
          "color": "#E5E5E5",
          "fontFamily": undefined,
          "fontSize": 36,
          "fontWeight": "700",
          "height": "16%",
          "letterSpacing": -0.2,
          "lineHeight": 56,
          "marginVertical": 0,
          "padding": 6,
          "paddingTop": 32,
          "textAlign": "center",
          "width": "100%",
          "zIndex": 1,
        }
      }
    >
      Scan QR
    </Text>
  </View>
  <View
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        {
          "flexDirection": "row",
        },
        {},
        undefined,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      action="primary"
      collapsable={false}
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      role="button"
      size="md"
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
            "backgroundColor": "#0077E6",
            "borderRadius": 4,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          {},
          {
            "backgroundColor": "#0077E6",
            "borderColor": "#4AA9FF",
          },
          {},
          {
            "height": 40,
            "paddingHorizontal": 20,
          },
          {},
          {
            "backgroundColor": "#0077E6",
            "borderColor": "#4AA9FF",
          },
          {},
          {
            "height": 40,
            "paddingHorizontal": 20,
          },
          {},
          {
            "margin": 30,
            "marginBottom": 40,
            "marginTop": 0,
          },
          undefined,
        ]
      }
      sx={
        {
          "margin": 30,
          "marginBottom": 40,
          "marginTop": 0,
        }
      }
      variant="solid"
    >
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          [
            undefined,
            {},
            {
              "color": "#525252",
              "fontFamily": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {},
            {
              "fontSize": 16,
              "lineHeight": 22,
            },
            undefined,
            {},
            {
              "color": "#FCFCFC",
            },
            {},
            {
              "color": "#FCFCFC",
              "fontWeight": "600",
            },
            {
              "color": "#005DB4",
            },
            {
              "color": "#FCFCFC",
            },
            {},
            {
              "color": "#FCFCFC",
            },
            undefined,
          ]
        }
      >
        Contacts
      </Text>
    </View>
  </View>
</View>
`;
